diff/windows/windows.go

```diff
@@ -332,6 +335,22 @@ func mountsToLayerAndParents(mounts []mount.Mount) (string, []string, error) {
 		return "", nil, err
 	}
 
+	if mnt.ReadOnly() {
```

Comment:
@TBBle given the changes in https://github.com/containerd/containerd/commit/d373ebc4de9a514af7d42c2edc980ec1d0868598, do you think the changes in this file are needed anymore?

Reply:
We can definitely get rid of _some_ of the changes (the ones that were caring about "bind", since we don't generate that type anymore), but the logic here might need revision as I _believe_ it was based on the old code where a View of a base layer still generated a scratch layer. (So I suspect it's not working currently here...)

It's possible that the old code worked fine. The case to watch out for is the one in that comment about `rootfs.CreateDiff`, as that code-path can pass a lower mount created by `sn.View(ctx, lowerKey, "")`, i.e. a read-only mount of a layer containing just a "Files" directory, and we need to ensure we trigger the base-layer exporter in hcsshim, hence the code in `mountsToLayerAndParents` that recognises such a mount and returns magic `""`, and the code in `mountPairToLayerStack` that recognises a lower of `""` and an upper with no parents, and returns just the upper.

This code-path is new with this PR, since until this PR, the Windows Snapshotter couldn't diff/export a base layer, it could only import them.

I'm not sure if any test code exercises that code-path, as it's mostly used by the checkpointing code that I suspect doesn't work in Windows for other reasons. But `ctr diff` also calls `rootfs.CreateDiff`, so I expect you could test the problematic code path by trying to `ctr diff` a base layer.

So my guess would be we still need _some_ changes in this file, since the existing code checked for the "new empty view" case assuming the snapshotter returned an empty list of mounts, apparently. But they can be simpler because we no longer have the case where we have to check for "bind" mount with parents, or a "windows-layer" mount without parents.

We definitely need to strip off the dummy sandbox on a non-base-layer View, otherwise the diff would be empty as we'd pass the dummy sandbox into hcsshim.

So  `mountsToLayerAndParents` only needs to keep the new `mnt.ReadOnly` code-path, and that can be simplified to something like:
```
	if mnt.ReadOnly() {
		if len(parentLayerPaths) == 0 {
			// A view without parents is a placeholder for "empty". We special-case this to simplify the caller.
			return "", nil, nil
		}
		// A view with parents points to a trivial scratch layer parented to the layer of interest, so strip it off.
		return parentLayerPaths[0], parentLayerPaths[1:], nil
	}
```

Then the logic in `mountPairToLayerStack` should still work; the main change this PR made was to look for that `""` as the lower layer path instead of an empty Mount array. I don't think it was ever possible to get an empty array there though (the Windows snapshotter always returns an error or a single Mount), so I suspect that code had never worked.

And a note for testing, `ctr diff` cannot diff an imported non-base layer, as HCS's diff export uses the sandbox.vhdx of the upper layer, and that only exists if a layer was created locally and committed. (Before the optimisation to skip sandbox.vhdx creation on imports, this case would produce an empty diff; now it should fail, again down in hcsshim).

So you'll have to create and populate the layers for testing locally.

Reply:
So, here's what I tried:

![Screenshot from 2023-04-03 17-46-30](https://user-images.githubusercontent.com/903160/229545194-aa1a9629-d2d8-4fd9-a517-36c8445bea02.png)

The ```testing``` snapshot was created by buildkit before trying to upload  ```.dockerignore``` which didn't exist, so it is empty. I then created a new active snapshot from it. Created a diff of the ```testing``` snapshot (has no parents), and one of the ```2``` snapshot.

Seems to do the right thing. This is with the changes you mentioned above.

Reply:
creating a diff of a layer that resulted from an image import seems to work as before. So does creating a diff of a child snapshot from that image.

If you see anything that I've missed, let me know.